Modules/sha2module.c

```diff
@@ -0,0 +1,789 @@
+#include "structmember.h"         // PyMemberDef
+#include "hashlib.h"
+
+/*[clinic input]
+module _sha2
+class SHA256Type "SHA256object *" "&PyType_Type"
+class SHA512Type "SHA512object *" "&PyType_Type"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=b5315a7b611c9afc]*/
+
+
+/* The SHA block sizes and maximum message digest sizes, in bytes */
+
+#define SHA256_BLOCKSIZE   64
+#define SHA256_DIGESTSIZE  32
+#define SHA512_BLOCKSIZE   128
+#define SHA512_DIGESTSIZE  64
+
+/* Our SHA2 implementations defer to the HACL* verified library. */
+
+#include "_hacl/Hacl_Streaming_SHA2.h"
+
+// TODO: transform .digestsize into a getter and get rid of the field?
+
+typedef struct {
+  PyObject_HEAD
+  int digestsize;
+  Hacl_Streaming_SHA2_state_sha2_256 *state;
+} SHA256object;
+
+typedef struct {
+    PyObject_HEAD
+    int digestsize;
+    Hacl_Streaming_SHA2_state_sha2_512 *state;
+} SHA512object;
+
+#include "clinic/sha2module.c.h"
+
+/* We shall use run-time type information in the remainder of this module to
+ * tell apart SHA2-224 and SHA2-256 */
+typedef struct {
+    PyTypeObject* sha224_type;
+    PyTypeObject* sha256_type;
+    PyTypeObject* sha384_type;
+    PyTypeObject* sha512_type;
+} sha2_state;
+
+static inline sha2_state*
+sha2_get_state(PyObject *module)
+{
+    void *state = PyModule_GetState(module);
+    assert(state != NULL);
+    return (sha2_state *)state;
+}
+
+static void SHA256copy(SHA256object *src, SHA256object *dest)
+{
+    dest->digestsize = src->digestsize;
+    dest->state = Hacl_Streaming_SHA2_copy_256(src->state);
+}
+
+static void SHA512copy(SHA512object *src, SHA512object *dest)
+{
+    dest->digestsize = src->digestsize;
+    dest->state = Hacl_Streaming_SHA2_copy_512(src->state);
+}
+
+static SHA256object *
+newSHA224object(sha2_state *state)
+{
+    SHA256object *sha = (SHA256object *)PyObject_GC_New(SHA256object,
+                                                  state->sha224_type);
+    PyObject_GC_Track(sha);
+    return sha;
+}
+
+static SHA256object *
+newSHA256object(sha2_state *state)
+{
+    SHA256object *sha = (SHA256object *)PyObject_GC_New(SHA256object,
+                                                  state->sha256_type);
+    PyObject_GC_Track(sha);
+    return sha;
+}
+
+static SHA512object *
+newSHA384object(sha2_state *st)
+{
+    SHA512object *sha = (SHA512object *)PyObject_GC_New(SHA512object, st->sha384_type);
+    PyObject_GC_Track(sha);
+    return sha;
+}
+
+static SHA512object *
+newSHA512object(sha2_state *st)
+{
+    SHA512object *sha = (SHA512object *)PyObject_GC_New(SHA512object, st->sha512_type);
+    PyObject_GC_Track(sha);
+    return sha;
+}
+
+/* Internal methods for our hash objects. */
+
+static int
+SHA2_traverse(PyObject *ptr, visitproc visit, void *arg)
+{
+    Py_VISIT(Py_TYPE(ptr));
+    return 0;
+}
+
+static void
+SHA256_dealloc(SHA256object *ptr)
+{
+    Hacl_Streaming_SHA2_free_256(ptr->state);
```

Comment:
Untrack first? Perhaps it does not matter.

Reply:
I don't think it matters here, the GIL is held. (long existing code)